From ff10c9950eca1eda967badea17539523ba69aac8 Mon Sep 17 00:00:00 2001
From: Tim Hartley <timothy.d.hartley@intel.com>
Date: Mon, 13 May 2013 22:22:00 +0000
Subject: Dalvik: Remove taken link to chaining cell after inlining

BZ: 107734

After inlining getters or setters, the taken link from the BB
to the chaining cell is still present. This removes it.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I64cc57d5cde51a9f9c8d463d4c7cbcbc4cc47308
Orig-MCG-Change-Id: If2b7354474457a01b954b739d74360440db8304c
Signed-off-by: Tim Hartley <timothy.d.hartley@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/InlineTransformation.cpp |   32 +++++++++++++++++++++++++++++++-
 1 files changed, 31 insertions(+), 1 deletions(-)
 mode change 100644 => 100755 vm/compiler/InlineTransformation.cpp

diff --git a/vm/compiler/InlineTransformation.cpp b/vm/compiler/InlineTransformation.cpp
old mode 100644
new mode 100755
index d7aa497..d021362
--- a/vm/compiler/InlineTransformation.cpp
+++ b/vm/compiler/InlineTransformation.cpp
@@ -376,7 +376,15 @@ static bool tryInlineVirtualCallsite(CompilationUnit *cUnit,
 }
 #endif
 
-bool dvmCompilerHandleInlining (CompilationUnit *cUnit, JitTranslationInfo *info, BasicBlock *bb, MIR *mir)
+/**
+ * @brief Help inline the instruction
+ * @param cUnit the CompilationUnit
+ * @param info the JitTranslationInfo
+ * @param bb the BasicBlock containing the instruction
+ * @param mir the instruction
+ * @return whether or not the inlining succeeded
+ */
+static bool handleInliningHelper (CompilationUnit *cUnit, JitTranslationInfo *info, BasicBlock *bb, MIR *mir)
 {
     Opcode opcode = mir->dalvikInsn.opcode;
     int flags = (int)dexGetFlagsFromOpcode(opcode);
@@ -546,6 +554,28 @@ bool dvmCompilerHandleInlining (CompilationUnit *cUnit, JitTranslationInfo *info
     return false;
 }
 
+bool dvmCompilerHandleInlining (CompilationUnit *cUnit, JitTranslationInfo *info, BasicBlock *bb, MIR *mir)
+{
+    bool res = handleInliningHelper (cUnit, info, bb, mir);
+
+    if (res == true)
+    {
+        //If we succeeded, we might have a singleton chaining cell to remove
+        BasicBlock *taken = bb->taken;
+
+        if (taken != 0 && taken->blockType == kChainingCellInvokeSingleton)
+        {
+            //Mark the taken as a dalvik byte code, it's no longer a chaining cell anymore
+            taken->blockType = kDalvikByteCode;
+
+            //Remove it, we don't need it anymore
+            bb->taken = 0;
+        }
+    }
+
+    return res;
+}
+
 void dvmCompilerInlineMIR(CompilationUnit *cUnit, JitTranslationInfo *info)
 {
     GrowableListIterator iterator;
-- 
1.7.4.1

